tools/generate.py

```diff
@@ -1,276 +1,404 @@
-            nbrNgramLine = len(text) - size
-            if nbrNgramLine > 0:
-                hyperLangNbrNgram[lang] += nbrNgramLine
-                for i in range(nbrNgramLine+1):
-                    ngram = text[i:i+size]
-                    hyperLangNgram[lang][ngram] += 1
-        print_status_line(size, lineNumber)
-        print(' done'.format(lineNumber))
-
-
-        print('Inserting ngrams of size {}...'.format(size))
-
-        table = TABLE_NGRAM + str(size)
-        
-        for lang, currentLangNgram in hyperLangNgram.items():
-            for ngram,hit in currentLangNgram.items():
-                freq = float(hit) / hyperLangNbrNgram[lang]
-
-                if lang in IDEOGRAM_LANGS:
-                    if freq > IDEOGRAM_NGRAM_FREQ_LIMIT:
-                        c.execute(
-                            INSERT_NGRAM % (table),
-                            (ngram,lang,hit,freq)
-                        )
-                else:
-                    if freq > NGRAM_FREQ_LIMIT:
-
-                        c.execute(
-                            INSERT_NGRAM % (table),
-                            (ngram,lang,hit,freq)
+class RawTatodetectDB:
+    """A sqlite database for tracking all n-grams counts and user
+    contributions in the Tatoeba corpus
+
+    It handles:
+    -   the occurrence counts of every n-gram found in the
+        Tatoeba corpus
+
+    -   the contribution scores of the Tatoeba users in every
+        language they have contributed to
+    """
+
+    def __init__(self, db_path: Path) -> None:
+        """
+        Parameters
+        ----------
+        db_path : Path
+            the location of the sqlite database
+        """
+        self._fp = db_path
+
+    def count(
+        self,
+        sentences_detailed_path: Path,
+        sentence_blacklist: list,
+        buffer_length: int = 5000000,
+    ) -> None:
+        """Count n-grams' occurrences for every Tatoeba language
+
+        A contribution score equal to sum(len(sentences))
+        is also computed for every language a user has contributed to.
+
+        All results are gradually saved into database tables in
+        order to mitigate the memory footprint of the process.
+
+        Parameters
+        ----------
+        sentences_detailed_path : Path
+            path of the 'sentences_detailed.csv' weekly dump file
+        sentence_blacklist : list
+            the integer ids of the sentences that are not taken
+            into account for this counting
+        buffer_length : int, optional
+            the maximum number of n-grams for which counts are kept in RAM,
+            by default 5,000,000.
+            Note that process speed mainly depends on the amount of data
+            written to disk and consequently increases with the buffer size.
+        """
+        # the former database is overwritten
+        if self._fp.exists():
+            print(f"Deleting former {self._fp.name} database")
+            self._fp.unlink()
```

Comment:
I think it would be better to remove the former database *after* the new database is successfully generated. Otherwise, if the new database generation fails for some reason, we end up with no database at all.

Reply:
Note that the [cronjob script](https://github.com/Tatoeba/tatoeba2/blob/a93f3331c95650edc89419e43117cfd476a3a1ab/docs/cron/regenerate-ngrams.sh) currently handles this already. The new database is written to a different file, and then only if the Python generation script does not error, the former database is replaced by the new (thanks to `set -e`).

Reply:
I chose to overwrite the former database to be sure not to add up the new counts to the former ones in the "raw database". Since the script is executed in a temp directory, there is no actual risk of data loss because of overwriting. So maybe the code could stay as it is?